internal/scheduler/metrics/metrics.go

```diff
@@ -250,14 +267,44 @@ func (m *Metrics) failedCategoryAndSubCategoryFromJob(ctx *armadacontext.Context
 	if run == nil {
 		return
 	}
+
 	category, message := errorTypeAndMessageFromError(ctx, jobRunErrorsByRunId[run.Id()])
+
+	i, ok := m.cachedRegexIndexFromErrorMessage(message)
+	if !ok {
+		i, ok = m.regexIndexFromErrorMessage(message)
+		if !ok {
+			// Use -1 to indicate that no regex matches.
+			i = -1
+		}
+		if m.matchedRegexIndexByErrorMessage != nil {
+			m.matchedRegexIndexByErrorMessage.Add(message, i)
+		}
+	}
+	if i != -1 {
+		subCategory = m.config.TrackedErrorRegexes[i]
+	}
+	return
+}
+
+func (m *Metrics) cachedRegexIndexFromErrorMessage(message string) (int, bool) {
+	if m.matchedRegexIndexByErrorMessage == nil {
+		return 0, false
+	}
+	i, ok := m.matchedRegexIndexByErrorMessage.Get(message)
+	if !ok {
+		return 0, false
+	}
+	return i.(int), true
+}
+
+func (m *Metrics) regexIndexFromErrorMessage(message string) (int, bool) {
```

Comment:
It occurs to me that it would be slightly cleaner to move all the interactions with the LRU cache into this function.

Reply:
That's prob. Let me try it.